tests/test_all.py

```diff
@@ -196,43 +198,65 @@ def test_glibc_present(auto_container):
 @pytest.mark.parametrize(
     "container_per_test", CONTAINERS_WITH_ZYPPER, indirect=True
 )
-def test_zypper_dup_works(container_per_test: ContainerData) -> None:
-    """Check that there are no packages installed that we wouldn't find in SLE
-    BCI repo by running :command:`zypper -n dup` and checking that there are no
-    conflicts or arch changes and we can update to the state in SLE_BCI repos.
-    Then validate that SLE_BCI provides all the packages that are afterwards
-    in the container as well except for the known intentional breakages
-    (sles-release, skelcd-EULA-bci).
-
-    As of 2023-05 the container and the SLE_BCI repositories are released independently
-    so we frequently get downgrades in this test. allow --allow-downgrade therefore
-    but still test that there wouldn't be conflicts with what is available in SLE_BCI.
+def test_no_downgrade_on_install(container_per_test: ContainerData) -> None:
+    """Check that we can install any additional pacakage in the container.
+
+    Check that installing any additional package would not cause a downgrade
+    of any package already installed in the container as that would throw
+    a question to the user and break the builds.
+
+    Also ensure that there is no package in the container that is not also
+    available in the BCI repository.
     """
     repo_name = "SLE_BCI"
     if OS_VERSION == "tumbleweed":
         repo_name = "repo-oss"
     if OS_VERSION == "basalt":
         repo_name = "repo-basalt"
 
-    container_per_test.connection.run_expect(
-        [0],
-        f"timeout 5m zypper -n dup --from {repo_name} -l "
-        "--no-allow-vendor-change --allow-downgrade --no-allow-arch-change",
-    )
+    conn = container_per_test.connection
 
-    searchresult = ET.fromstring(
-        container_per_test.connection.run_expect(
-            [0], "zypper -x -n search -t package -v -i '*'"
-        ).stdout
+    conn.run_expect([0], f"timeout 2m zypper ref")
+    system_solv = json.loads(
+        conn.check_output("dumpsolv -j /var/cache/zypp/solv/@System/solv")
     )
+    bci_solv = json.loads(
+        conn.check_output(f"dumpsolv -j /var/cache/zypp/solv/{repo_name}/solv")
+    )
+    installed_pkgs = {
+        solvable["solvable:name"]: solvable["solvable:evr"]
+        for solvable in system_solv["repositories"][0]["solvables"]
+    }
+    bci_pkgs = {}
+    for solvable in bci_solv["repositories"][0]["solvables"]:
+        bci_pkgs[solvable["solvable:name"]] = solvable["solvable:evr"]
+        if solvable["solvable:name"] in installed_pkgs:
+            continue
+        for req in solvable.get("solvable:requires", ()):
+            # Skip boolean dependencies or unversioned ones
+            if "(" in req or " = " not in req:
```

Comment:
This will also skip virtual requires like `pkgconfig(libfoo) = 1.2.3`.

Reply:
@dcermak yeah, I didn't want to implement provides/requires matching logic. so far it was finding the essential issues without that (as typically people do `Requires: %name = %version-%release` somewhere. we can always improve that later if it is necessary to find more issues (but already now it finds too many :/ )